include/cuco/detail/static_map/kernels.cuh

```diff
@@ -108,4 +108,105 @@ __global__ void insert_or_apply(InputIt first, cuco::detail::index_type n, Op op
   }
 }
 
+template <int32_t CGSize,
+          int32_t BlockSize,
+          typename SharedMapRefType,
+          typename InputIt,
+          typename Op,
+          typename Ref>
+CUCO_KERNEL __launch_bounds__(BlockSize) void insert_or_apply_shmem(
+  InputIt first,
+  cuco::detail::index_type n,
+  Op op,
+  Ref ref,
+  typename SharedMapRefType::extent_type window_extent)
+{
+  namespace cg     = cooperative_groups;
+  using Key        = typename Ref::key_type;
+  using Value      = typename Ref::mapped_type;
+  using value_type = typename std::iterator_traits<InputIt>::value_type;
+
+  auto const block       = cg::this_thread_block();
+  auto const thread_idx  = block.thread_rank();
+  auto const loop_stride = cuco::detail::grid_stride() / CGSize;
+  auto idx               = cuco::detail::global_thread_id() / CGSize;
+
+  auto warp                  = cg::tiled_partition<32>(block);
+  auto const warp_thread_idx = warp.thread_rank();
+
+  // Shared map initialization
+  __shared__ typename SharedMapRefType::window_type windows[window_extent.value()];
+  auto storage           = SharedMapRefType::storage_ref_type(window_extent, windows);
+  auto const num_windows = storage.num_windows();
+
+  __shared__ int32_t block_cardinality;
+  if (thread_idx == 0) block_cardinality = 0;
+  block.sync();
+
+  cuda::atomic_ref<int32_t, cuda::thread_scope_block> cardinality_counter(block_cardinality);
```

Comment:
Instead of placing a plain integer in shmem and then wrapping it with `atomic_ref`, you could use a `__shared__ cuda::atomic<...> block_cardinality`. Note that you have to properly initialize the object using placement-`new`.